### PR TITLE
Docs: using targetted styles, not components to style MDX

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: Our MDX components no longer work for pure HTML in MDX, in particular our `Introduction.mdx`.

**Current:** 
image

**6.x**

image

The root cause here is a breaking change in MDX2: https://mdxjs.com/migrating/v2/#update-mdx-content

If you put down (as we do in `Introduction.mdx`):

```html
<div className="subheading">Configure</div>
```

It will no longer be replaced by the component you specify for `<div>` to MDX. Currently we pass a "styled div" that (in the case of the `<div>`) does a reset and sets fonts etc. This will not work in MDX2.

There are a couple of problems:

1. Our introduction's styles are messed up.
2. Any user that has relied on our styling for raw HTML in MDX is going to have an issue.

## What I did

@JReinhold and I have prototyped a solution using targetted styles on the `DocsContent` wrapper. Basically for the above, we use:

```css
div:not(.sb-story div) { /* styles */ }
```

To target styles at all `<div>` elements that *are not children of `.sb-story`* (we will attach that style to the output of the `<Story/>` block). That ensures the styles apply inside the MDX but not inside stories.

However, there are some issues with this approach:

1. Some of the components do more than just styles, primarily the `<code>`. In this case, we won't be able achieve full parity.

2. It seems like MDX is inserting some extra `<p>` elements that change the structure of the generated HTML anyway. I'm not sure if this is a bug, but this will likely mess people up too.

3. The specificity of our styles change as a result -- before it was `.css-2ltd27` (some emotion class), now it it `.css-2ltd27 div:not(.sb-story div)` which is a higher specificity, which is going to be a breaking change (in particular we need to retarget the inline styles in `Introduction.mdx` to ensure they override the reset). 

4. The styles will target other elements inside components the user uses in MDX, such as our doc blocks, or 3rd party "blocks".

## Other options

Some other alternatives we've considered:

**Alternative 1**: Simply change the introduction to look nice (by adding more styles) and make raw HTML in MDX completely unstyled.

**Alternative 2**: As above, but include a very basic set of styles (e.g fonts, line-height) on the wrapper element, let it cascade, then *reset it* on the `<Story/>` element.

## Other notes

- Due to 2-3, I believe it's not going to be possible for this not to be a breaking change to some degree for users using raw HTML. We might take the opportunity to simplify things a little (i.e. do alternative 2).

### CODE REVIEW

**Review by ndelangen — APPROVED** [review]

**Comment by JReinhold** [comment]
@tom There's one minor issue with our opt-out approach, namely style inheritance. [Some styles are inherited](https://stackoverflow.com/questions/5612302/which-css-properties-are-inherited) by parents even if they are not set.
Eg. adding `color: green` to a root div will cause all descendants to have `color: green`, (where as the same thing is not true for `border`, not all descendants will have a border just because an ancestor has one).

With our global selector, this means that styles can get inherited even through an `<Unstyled>` or `<Canvas>` component.

Examples:

```
<Unstyled>     <-- blocks our global styles, this is great, it works
  <div>
    {'this is text'}
    <Story of={StoriesMeta.Undefined} meta={StoriesMeta}/>
  </div>
</Unstyled>

<div>     <-- gets styled by our global styles
  {'this is text'}
  <Unstyled>     <-- blocks global styles for all descendants, but not the parent div 
    <Story of={StoriesMeta.Undefined} meta={StoriesMeta}/>     <-- will inherit the global styles from the div, even though it doesn't have them directly
  </Unstyled>
</div>
```

if our global styles styled `div` with `color: green`, the above would output this:
image

Solutions:
1. This might be a minor thing, that we could "ignore". I expect it would mostly be an issue with inheriting our font family and size.
2. We could document that `Unstyled`, `Canvas` and `Story` will not behave as intended unless they are added to the root markdown
3. I've tried with the new `:has` selector to select parents, so that the second example would also select the ancestor div and make it unstyled, and it works, but I'm afraid this would be unexpected behavior for the user? [It is also still not supported in Firefox](https://caniuse.com/css-has) (it will break the whole global style, not just ignore the `:has` part).

**Comment by JReinhold** [comment]
Also, I looked into the few components that still used the components from `@storybook/components`, and it wasn't straight forward to yank them out. I still think it's important that we do it so the styles don't go out of sync, but in the interest  of time I've decided to postpone that work.

File: code/ui/blocks/src/blocks/Unstyled.mdx [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { Unstyled } from './Unstyled.tsx';
```
Comment (JReinhold) [inline code/ui/blocks/src/blocks/Unstyled.mdx:1]: @jonniebigodes if you have feedback for this documentation, let me know.

File: code/ui/blocks/src/components/DocsPage.tsx [hunk 2]
```diff
@@ -1,9 +1,32 @@
...
+ * 4. .sb-unstyled is an escape hatch that allows the user to opt-out of the default styles
+ *    by wrapping their content in an element with this class: <Unstyled />
+ * 5. the other UNSTYLED_SELECTORS are elements we don't want the styles to bleed into, like canvas, story and source blocks.
+ */
+const UNSTYLED_SELECTORS = [
+  '.sb-unstyled',
+  '.sbdocs-preview',
+  '.sb-story',
+  '.docblock-source',
+  '.sb-anchor',
```
Comment (JReinhold) [inline code/ui/blocks/src/components/DocsPage.tsx:25]: It's up for discussion how the responsibility should be divided here.

This selector could also just be `sb-unstyled`, and then it was the responsibility of any of the components (`Story`, `Canvas`, etc.) to add that class to their elements.

Who is responsible for unstyling elements, the styler (here) or the element?

File: code/ui/blocks/src/components/DocsPage.tsx [hunk 3]
```diff
@@ -40,9 +63,374 @@ export const Subtitle = styled.h2(withReset, ({ theme }) => ({
   color: transparentize(0.25, theme.color.defaultText),
 }));
```
Comment (JReinhold) [inline code/ui/blocks/src/components/DocsPage.tsx:65]: The following style are copied from all the components in https://github.com/storybookjs/storybook/tree/tom/sb-996-resolve-css-issue-with-introduction/code/ui/components/src/typography/elements

File: code/ui/blocks/src/components/DocsPage.tsx (hunk 2 above)
Comment (tmeasday) [reply code/ui/blocks/src/components/DocsPage.tsx:25]: I'm open to whatever but my instinct was the other way round (`.sb-unstyled`)

**Comment by tmeasday** [comment]
@JReinhold I get it, I think that limitation makes sense and is an issue in the 6.x implementation as well, right? I think we can live with it for sure.

File: code/ui/blocks/src/components/DocsPage.stories.tsx [hunk 4]
```diff
@@ -95,3 +96,28 @@ export const Markdown = () => (
...
+    <DocsPageWrapper>
+      <h1>Heading 1</h1>
+      <h2>Heading 2</h2>
+      <a>A tag</a>
+      <pre>pre tag</pre>
+      <div>
+        <div>Div</div>
+        <a>Nested A tag</a>
+      </div>
+      <div className="sb-unstyled" style={{ border: '2px solid red' }}>
```
Comment (tmeasday) [inline code/ui/blocks/src/components/DocsPage.stories.tsx:112]: Why not use the `Unstyled` component here?

File: code/ui/blocks/src/components/DocsPage.tsx (hunk 2 above)
Comment (JReinhold) [reply code/ui/blocks/src/components/DocsPage.tsx:25]: Let's keep it as-is then and revisit if we get wiser.

**Comment by JReinhold** [comment]
> is an issue in the 6.x implementation as well, right?

Huh, you must be right, I hadn't thought about that at all. No problem then.

**Review by tmeasday — COMMENTED** [review]
Awesome @JReinhold! Can we add some notes about breaking change to the `MIGRATION.md` (cc @shilman / @jonniebigodes we should maybe mention in MDX2 docs too):

- styles leak into non-story components if you don't wrap in `<Unstlyed>`
- `<Code>` needs to be imported now

**Comment by tmeasday** [comment]
@JReinhold I can't approve as I opened this PR, but I'm happy with this.

File: code/ui/blocks/src/components/DocsPage.stories.tsx (hunk 4 above)
Comment (JReinhold) [reply code/ui/blocks/src/components/DocsPage.stories.tsx:112]: You're right, I did this before I built the component.

**Comment by JReinhold** [comment]
@tmeasday I've added some entries to the [migration doc, please check it out](https://github.com/storybookjs/storybook/pull/19958/files#diff-c1603ee6a674bb75255aa448057630506153e363297fd58668e6679f10909bc6).

> * `<Code>` needs to be imported now

This is actually not possible currently, because the component that replaces `code` isn't exported to the user. I've tried:

- using `Code` from `@storybook/components`, but that it doesn't look quite the same, the font is not the same size at least.
- using `Source` from `@storybook/blocks` but that doesn't work without a Meta tag, so I don't see that as an option
- exporting and using `Source` from `@storybook/blocks/components` but it's very unergonomic.

As you can see from my migration notes, I think the use cases for defining a code block with `<code>` are diminishing with MDX2's better support for code fences, so maybe we don't need a mitigation step for this at all? It may only be when a user wants to add custom classes/ids to their code blocks that it would be useful to them.

File: code/ui/blocks/src/components/DocsPage.tsx (hunk 2 above)
Comment (JReinhold) [reply code/ui/blocks/src/components/DocsPage.tsx]: Narrator: He did get wiser. https://github.com/storybookjs/storybook/pull/21150

**Pull request merged**